tests/profiles/ufo_sources_test.py

```diff
@@ -67,23 +69,23 @@ def test_check_recommended_fields(empty_ufo_font):
     ufo.info.copyright = "Test"
     ufo.info.openTypeOS2Panose = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
 
-    assert_PASS(check(ufo),
+    assert_PASS(check({}, {"ufo_font": ufo}),
                 'with an almost empty UFO.')
 
 
 def test_check_unnecessary_fields(empty_ufo_font):
-    from fontbakery.profiles.ufo_sources import com_daltonmaag_check_unnecessary_fields as check
+    check = CheckTester(ufo_sources_profile, "com.daltonmaag/check/ufo-unnecessary-fields")
     ufo, _ = empty_ufo_font
 
-    assert_PASS(check(ufo),
+    assert_PASS(check({}, {"ufo_font": ufo}),
```

Comment:
Why `{}` here? Checks can operate on UFO files.

Reply:
But CheckTester doesn’t have the logic to turn “ufo” arguments into the right type; so we have to process the condition for it.

Reply:
I think we should add that logic, then. Because the checking should mimic actual usage of the check as much as possible.

Reply:
I think I would prefer to go the other way: the call to the `CheckTester` object should *always* take a kwargs dictionary which maps variable names in the check function to their values.

So instead of

```
check(good_ttFont) # We do complex logic to guess that this gets fed to the check's "ttFont" parameter
```

we explicitly say

```
check(ttFont=good_ttFont)
```

What do you think?

Reply:
That might be acceptable but not ideal because it diverges from my original intentions. My hope was to transform the CheckTester class into the key element in an actual user-friendly "FontBakery API", so it should be smart to chew up anything that fontbakery's command line typically accepts.

Reply:
Okay. But the command line only accepts file names - that has limited use in testing. I suppose we *could* go that way, but it would mean writing out every test font to disk before running each test. It might ultimately simplify things because we could then use CheckRunner directly instead of having to fake dependencies etc.

I think we probably need to think about this some more. Is CheckTester there to test low-level functionality (i.e. the check functions themselves) or high-level functionality (i.e. the entire operation of fontbakery)? I really think we want to operate our regression tests at a granular level, testing the actual check function. But that is very different from a "smart" FontBakery API.

Reply:
Take a look at PR #3444 in which I added the logic for chewing up GlyphsApp files. Since we've just had a v0.8.2 release, I will merge that PR now in the hope we can facilitate work here and knowing that we'll likely have a couple weeks until FB 0.8.3, so it seems to be enough time to address any pending issues on that PR. (I think the only actual issue is that the current check-ids are not what we'll actually use once we have check polymorphism)

Reply:
I'm really sorry; I still don't understand the implications of this. When you say:

> The way these code-tests are written is meant to imitate a user performing checks on fonts. The implementation of the code-tests should look like normal usage of the tool.

The obvious question is "at what level"? When a user performs checks on a font using `fontbakery` normally, the "parameters" are a list of filenames. Do you mean that `CheckTester` should also take a list of filenames?

If so, we will need to rewrite a large number of the tests, as they assume that you can inject conditions (often modified `ttFont` objects) into the test.

Reply:
Well... you're right. The code-tests do not work only on lists of filepaths. They can indeed also accept in-memory objects such as `ttFont`. (doing so facilitates the implementation but still keeps the mental model of "this is what a user might feed into a check")

And you're also correct about passing overrides to conditions. That's our way of fabricating the testing scenarios.

But it is not true that `"they assume that the parameters to the check(...) call are the same as the parameters to the underlying check function."`

In the past, our code-tests had to do it and they were more error prone as there was much more boilerplate code for explicitly computing the conditions. Then we created `CheckTester` with the purpose of performing such condition resolution automatically and transparently, so that the actual code-tests would be focused on the input conditions and expected check results: "If I give you this font, with this specific bit set, be sure that the check detects the problem and reports a **WARN**"

Indeed, in some cases we pass not only the input object (that used to be only font files but now can be anything), but we also pass overrides to certain conditions. But that's not mandatory - you don't have to pass things just because it is in the check function signature, CheckTester can do that for us. Condition overrides are just an optional mechanism exclusively used to fake testing scenarios because it would be worse to provide sample files for all possible bad-font scenarios.

The problem with the current approach in some portions of this PR is that when it passes an empty list for a check and then uses the condition overrides for feeding the main things under checking it is breaking the abstraction because it no longer seem to clearly express what a user would do: no user runs fontbakery against an empty list.

I understand that the end-result will still be the same, as stuff is being fed via the overrides, but it diverges from the established style of closely mimicking actual usage.

Reply:
Here's a before/after comparison of the code-test styles back from when the `CheckTester` class and this concept was introduced in our code-testing practices:

![photo_2020-09-03_12-10-30](https://user-images.githubusercontent.com/213676/132068250-c8639e29-f904-470e-8f71-88781acdc845.jpg)

Reply:
**note:** The class was called `TestingContext` instead of `CheckTester` back then, but the implementation was almost exactly the same as we still currently have.